[ClangImporter] Fix NS_CLOSED_ENUM to validate raw values

an0 opened the pull request:
## Summary

Fixes #85701 - `NS_CLOSED_ENUM` now properly validates raw values in `init?(rawValue:)`, returning `nil` for invalid values instead of creating invalid enum instances that crash later.

## Problem

When importing C enums marked with `NS_CLOSED_ENUM` (which map to `@frozen` Swift enums), the synthesized `init?(rawValue:)` used `Builtin.reinterpretCast` without validation. This allowed creating enum instances with invalid raw values that would crash with "Fatal error: unexpected enum case" when used in switch statements.

## Solution

- **For `@frozen` enums (`NS_CLOSED_ENUM`):** Generate a switch statement that validates the raw value against all declared cases, returning `nil` for invalid values
- **For non-frozen enums (`NS_ENUM`):** Preserve existing `reinterpretCast` behavior for C compatibility

## Implementation

Modified `synthesizeEnumRawValueConstructorBody()` in `lib/ClangImporter/SwiftDeclSynthesizer.cpp` to detect frozen enums and generate switch-based validation, following the same pattern used for native Swift enums in `DerivedConformanceRawRepresentable.cpp`.

## Testing

Added comprehensive execution tests that verify:
- Valid raw values correctly initialize enum instances
- Invalid raw values return `nil` (not crash)
- Non-contiguous and negative raw values handled correctly
- Open enums still accept arbitrary values for C compatibility

## Performance

Matches native Swift enums.

an0 committed:
[ClangImporter] Add tests for NS_CLOSED_ENUM raw value validation

Add execution tests verifying that NS_CLOSED_ENUM (imported as @frozen)
properly validates raw values in init?(rawValue:), returning nil for
invalid values instead of creating poison enum instances.

These tests currently fail, demonstrating issue #85701 where invalid
raw values are accepted and later crash in switch statements.

an0 committed:
[ClangImporter] Fix NS_CLOSED_ENUM to validate raw values

Generate switch-based validation for @frozen enum init?(rawValue:)
instead of using unchecked reinterpretCast. This ensures only declared
enum cases are accepted, returning nil for invalid raw values.

For non-frozen enums (NS_ENUM), preserve existing reinterpretCast
behavior for C compatibility.

Fixes #85701

j-hui commented on test/ClangImporter/enum-closed-raw-value.swift:
```diff
@@ -0,0 +1,84 @@
+// RUN: %empty-directory(%t)
+// RUN: %target-build-swift %s -import-objc-header %S/Inputs/enum-closed-raw-value.h -o %t/a.out
```
Can you please test this both with and without C++ interop?

j-hui commented on lib/ClangImporter/SwiftDeclSynthesizer.cpp:
```diff
@@ -1290,11 +1335,92 @@ synthesizeEnumRawValueConstructorBody(AbstractFunctionDecl *afd,
+
+    auto body = BraceStmt::create(ctx, SourceLoc(), {switchStmt}, SourceLoc(),
+                                  /*implicit*/ true);
+    return {body, /*isTypeChecked=*/false};
```
Can you please elaborate on why `isTypeChecked` is returned as `false` here?

j-hui reviewed (commented):
Thank you for your contribution! The compiler changes look good to me, aside from a question and a nit. I would like the test to be migrated to use `StdlibUnittest` though.

j-hui commented:
@swift-ci please test

an0 commented:
@j-hui I've addressed all the review feedback:
  - ✅ Converted tests to StdlibUnittest framework with `expectEqual()`, `expectNil()`, etc.
  - ✅ Added C++ interop test variant (two RUN lines)
  - ✅ Removed unnecessary else branch in SwiftDeclSynthesizer.cpp
  - ✅ Added comments explaining `isTypeChecked` reasoning for both code paths

AdamCmiel commented on lib/ClangImporter/SwiftDeclSynthesizer.cpp:
```diff
@@ -1290,10 +1335,95 @@ synthesizeEnumRawValueConstructorBody(AbstractFunctionDecl *afd,
+      auto *caseBody = BraceStmt::create(ctx, SourceLoc(), {caseAssign},
+                                         SourceLoc(), /*implicit*/ true);
+
+      auto *caseStmt = CaseStmt::createImplicit(ctx, CaseParentKind::Switch,
```
Great contribution! You could save a few bytes by generating one case statement with each known case as a series of labels, as their bodies are the same (reinterpret cast)

equivalent to the source:
```
switch rawValue {
case .first, .second, third:
  reinterpretCast
default:
  return nil
}
```

function merging may consolidate these blocks on the back end but better not to emit them in the first place

j-hui commented on lib/ClangImporter/SwiftDeclSynthesizer.cpp:
```diff
@@ -1247,7 +1247,52 @@ SwiftDeclSynthesizer::makeIndirectFieldAccessors(
+///       self = Builtin.reinterpretCast(rawValue)
+///     ...
+///     default:
+///       return nil
```
nit: this comment should be updated to reflect the most recent change:

```suggestion
///   init?(rawValue: RawType) {
///     switch rawValue {
///     case <value1>, <value2>, ...:
///       self = Builtin.reinterpretCast(rawValue)
///     default:
///       return nil
```

j-hui reviewed (approved):
Thanks for the ping @an0 , this looks really great!

I just have a couple of nits that I would still like you to address before merging, but approve of this PR overall.

Feel free to ping me to help you merge if you don't have permissions to do so yourself.

an0 commented on lib/ClangImporter/SwiftDeclSynthesizer.cpp:
```diff
@@ -1290,10 +1335,73 @@ synthesizeEnumRawValueConstructorBody(AbstractFunctionDecl *afd,
+
+  if (isFrozen) {
+    // For frozen enums, generate a switch statement to validate the raw value
+    SmallVector<CaseStmt *, 8> cases;
```
> Feel free to ping me to help you merge if you don't have permissions to do so yourself.

I think you are right, I don't see any button to merge it, so I need your help.

an0 commented:
> Does this resolve https://github.com/swiftlang/swift/issues/58272?

Yes. You can mark your report as dup of #85701
